From 2024f63b3f4be5e591117fefbcae4320db3041d7 Mon Sep 17 00:00:00 2001
From: Yahan Lu <yahan@iscas.ac.cn>
Date: Fri, 16 Apr 2021 12:35:49 +0800
Subject: [PATCH] [riscv64][wasm] Fix OSR shadow stack violation

Port 06a2c2e0c07cb29ae09d3f0375c3d03b25be3433

Change-Id: I23cbeabc14562168b5060f435beb77778dcfccdc
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/2831171
Commit-Queue: Brice Dobry <brice.dobry@futurewei.com>
Reviewed-by: Brice Dobry <brice.dobry@futurewei.com>
Cr-Commit-Position: refs/heads/master@{#74011}
---
 src/builtins/riscv64/builtins-riscv64.cc              | 4 ++++
 src/wasm/baseline/riscv64/liftoff-assembler-riscv64.h | 3 +++
 2 files changed, 7 insertions(+)

diff --git a/src/builtins/riscv64/builtins-riscv64.cc b/src/builtins/riscv64/builtins-riscv64.cc
index 15e0aab74f3..5c3dfb36817 100644
--- a/src/builtins/riscv64/builtins-riscv64.cc
+++ b/src/builtins/riscv64/builtins-riscv64.cc
@@ -2932,6 +2932,10 @@ void Builtins::Generate_GenericJSToWasmWrapper(MacroAssembler* masm) {
   __ Trap();
 }
 
+void Builtins::Generate_WasmOnStackReplace(MacroAssembler* masm) {
+  // Only needed on x64.
+  __ Trap();
+}
 namespace {
 
 int AddressOffset(ExternalReference ref0, ExternalReference ref1) {
diff --git a/src/wasm/baseline/riscv64/liftoff-assembler-riscv64.h b/src/wasm/baseline/riscv64/liftoff-assembler-riscv64.h
index ecb97aa32a0..d2039253435 100644
--- a/src/wasm/baseline/riscv64/liftoff-assembler-riscv64.h
+++ b/src/wasm/baseline/riscv64/liftoff-assembler-riscv64.h
@@ -408,6 +408,8 @@ void LiftoffAssembler::SpillInstance(Register instance) {
   Sd(instance, liftoff::GetInstanceOperand());
 }
 
+void LiftoffAssembler::ResetOSRTarget() {}
+
 void LiftoffAssembler::FillInstanceInto(Register dst) {
   Ld(dst, liftoff::GetInstanceOperand());
 }
@@ -2586,6 +2588,7 @@ void LiftoffAssembler::DeallocateStackSlot(uint32_t size) {
   Add64(sp, sp, Operand(size));
 }
 
+void LiftoffAssembler::MaybeOSR() {}
 
 void LiftoffStackSlots::Construct(int param_slots) {
   DCHECK_LT(0, slots_.size());
-- 
2.35.1

